src/autoscaler/operator/applicationsyncer.go

```diff
@@ -0,0 +1,46 @@
+package operator
+
+import (
+	"autoscaler/cf"
+	"autoscaler/db"
+
+	"code.cloudfoundry.org/lager"
+)
+
+type ApplicationSynchronizer struct {
+	cfClient cf.CFClient
+	policyDb db.PolicyDB
+	logger   lager.Logger
+}
+
+func NewApplicationSynchronizer(cfClient cf.CFClient, policyDb db.PolicyDB, logger lager.Logger) *ApplicationSynchronizer {
+	return &ApplicationSynchronizer{
+		policyDb: policyDb,
+		cfClient: cfClient,
+		logger:   logger,
+	}
+}
+
+func (as ApplicationSynchronizer) Operate() {
+	as.logger.Debug("deleting non-existent application details")
+	// Get all the application details from policyDB
+	appIds, err := as.policyDb.GetAppIds()
+	if err != nil {
+		as.logger.Error("failed-to-get-apps", err)
+		return
+	}
+	// For each app check if they really exist or not via CC api call
+	for appID := range appIds {
+		_, err = as.cfClient.GetApp(appID)
```

Comment:
There are multiple possible errors in GetApp() calls
For example,  if the status code = 200 OK,  an error returned: 
https://github.com/cloudfoundry-incubator/app-autoscaler/blob/9f93c7c2b4ae3091b7b409b855755555c74a6fd6/src/autoscaler/cf/app.go#L41-L45
With the code above,  what will happen when the CC api endpoint is temporarily unavailable  ? All GetAPP() may fail with error, then all policy might be deleted.    We met such a situation in daily operations ..

Technically, we can only delete App with the following response:
RESPONSE: [2018-09-05T10:55:18+08:00]
HTTP/1.1 404 Not Found
{
  "description": "The app could not be found: xxxxx",
  "error_code": "CF-AppNotFound",
  "code": 100004
}
Otherwise, we may false delete a valid policy.

Reply:
@cdlliuy  PR updated after incorporating the review comments